lightstep/sdk/metric/producer.go

```diff
@@ -0,0 +1,120 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metric // import "github.com/lightstep/otel-launcher-go/lightstep/sdk/metric"
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/lightstep/otel-launcher-go/lightstep/sdk/metric/data"
+	"github.com/lightstep/otel-launcher-go/lightstep/sdk/metric/internal/asyncstate"
+)
+
+// providerProducer is the binding between the MeterProvider and the
+// Reader.  This is the Producer instance that is passed to Register()
+// for each Reader.
+type providerProducer struct {
+	lock        sync.Mutex
+	provider    *MeterProvider
+	pipe        int
+	lastCollect time.Time
+}
+
+// producerFor returns the new Producer for calling Register.
+func (mp *MeterProvider) producerFor(pipe int) Producer {
+	return &providerProducer{
+		provider:    mp,
+		pipe:        pipe,
+		lastCollect: mp.startTime,
+	}
+}
+
+// Produce runs collection and produces a new metrics data object.
+func (pp *providerProducer) Produce(inout *data.Metrics) data.Metrics {
+	ordered := pp.provider.getOrdered()
+
+	// Note: the Last time is only used in delta-temporality
+	// scenarios.  This lock protects the only stateful change in
+	// `pp` but does not prevent concurrent collection.  If a
+	// delta-temporality reader were to call Produce
+	// concurrently, the results would be be recorded with
+	// non-overlapping timestamps but would have been collected in
+	// an overlapping way.
+	pp.lock.Lock()
+	lastTime := pp.lastCollect
+	nowTime := time.Now()
+	pp.lastCollect = nowTime
+	pp.lock.Unlock()
+
+	var output data.Metrics
+	if inout != nil {
+		inout.Reset()
+		output = *inout
+	}
+
+	output.Resource = pp.provider.cfg.res
+
+	sequence := data.Sequence{
+		Start: pp.provider.startTime,
+		Last:  lastTime,
+		Now:   nowTime,
+	}
+
+	// TODO: Add a timeout to the context.
+	ctx := context.Background()
+
+	for _, meter := range ordered {
+		meter.collectFor(
+			ctx,
+			pp.pipe,
+			sequence,
+			&output,
+		)
+	}
+
+	return output
+}
+
+// collectFor collects from a single meter.
+func (m *meter) collectFor(ctx context.Context, pipe int, seq data.Sequence, output *data.Metrics) {
+	// Use m.lock to briefly access the current lists: syncInsts, asyncInsts, callbacks
+	m.lock.Lock()
+	syncInsts := m.syncInsts
+	asyncInsts := m.asyncInsts
+	callbacks := m.callbacks
+	m.lock.Unlock()
```

Comment:
By releasing these locks, we allow new instruments and callbacks to be registered while collection happens. The items themselves are synchronized, and the slices are only appended to, so shallow copies are safe. If new instruments and callbacks are registered while this collection happens, they simply will not collect and any activity they experience concurrently will be registered on the next collection by this reader.

Reply:
Ah, i was under the assumption that because these were slices, any changes to the collections would affect this collection cycle for the reader. Thank you.